### ISSUE TITLE
[9.3.x] contracts with associated elements are throwing PHP backtrace if cronjob is executed

### ISSUE THREAD

**Issue by floppyist** [glpi-project/glpi#5117]
If a contracts has at least one element associated, the cronjob throws a PHP backtrace.

1. Create a contract which runs into the cancellation period a day before
2. Add one or more elements to the contract
3. Goto "automatic actions"; choose "contract"
4. reset "Last Run"
4. Wait until cronjob is executed

**Logs**
```
Toolbox::userErrorHandlerNormal() in <glpi_path>/inc/toolbox.class.php line 657
  *** PHP Notice(8): Undefined index: glpiactiveentities
  Backtrace :
  inc/dbutils.class.php:630                          
  inc/db.function.php:684                            DbUtils->getEntitiesRestrictCriteria()
  inc/commondbrelation.class.php:1732                getEntitiesRestrictCriteria()
  inc/commondbrelation.class.php:1752                CommonDBRelation::getTypeItemsQueryParams()
  inc/contract_item.class.php:206                    CommonDBRelation::getTypeItems()
  inc/contract.class.php:1252                        Contract_Item::getItemsForContract()
  :                                                  Contract::cronContract()
  inc/crontask.class.php:837                         call_user_func()
  front/cron.php:83                                  CronTask::launch()
 Toolbox::userErrorHandlerNormal() <glpi_path>/inc/toolbox.class.php line 657
  *** PHP Notice(8): Undefined index: glpiparententities
  Backtrace :
  inc/dbutils.class.php:652                          
  inc/db.function.php:684                            DbUtils->getEntitiesRestrictCriteria()
  inc/commondbrelation.class.php:1732                getEntitiesRestrictCriteria()
  inc/commondbrelation.class.php:1752                CommonDBRelation::getTypeItemsQueryParams()
  inc/contract_item.class.php:206                    CommonDBRelation::getTypeItems()
  inc/contract.class.php:1252                        Contract_Item::getItemsForContract()
  :                                                  Contract::cronContract()
  inc/crontask.class.php:837                         call_user_func()
  front/cron.php:83                                  CronTask::launch()
```

**Additional context**
This ticket is related to #5107 

I tested with existing elements (before migration) and also with new ones. The backtrace is always shown.

**Comment by floppyist** [comment]
Is there any update? We almost get a daily warning sent by cron because of this issue.

Thanks.

**Comment by cedric-anne** [comment]
Hi,

#5250 should fix your issue.

Regards

**Comment by floppyist** [comment]
Thanks for the quick feedback. But the fix is only available for 9.4. We testet the fix on 9.3.1 an we got the following error message after cronjob was executed: 

> 	PHP Parse error: syntax error, unexpected 'if' (T_IF) in <document_root>/inc/dbutils.class.php on line 613 

We have to stick with 9.3.1 currently, maybe you can help me again?

Regards

**Comment by trasher** [comment]
You should be at least on 9.3.3.

Next release is 9.4.0, you can wait for it to be released, or you can try to apply the fix on your local instance.

**Comment by floppyist** [comment]
We manually done the fix in 9.3.3. It works. Thank you guys.

Regards
